src/SOS/SOS.Extensions/Clrma/ClrmaServiceWrapper.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Diagnostics.DebugServices;
+using Microsoft.Diagnostics.Runtime.Utilities;
+using SOS.Hosting;
+using SOS.Hosting.DbgEng.Interop;
+
+namespace SOS.Extensions.Clrma
+{
+    public sealed class ClrmaServiceWrapper : COMCallableIUnknown
+    {
+        public static readonly Guid IID_ICLRMAService = new("1FCF4C14-60C1-44E6-84ED-20506EF3DC60");
+
+        public const int E_BOUNDS = unchecked((int)0x8000000B);
+        public const uint DEBUG_ANY_ID = uint.MaxValue;
+
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ServiceWrapper _serviceWrapper;
+        private ICrashInfoService _crashInfoService;
+
+        public ClrmaServiceWrapper(ITarget target, IServiceProvider serviceProvider, ServiceWrapper serviceWrapper)
+        {
+            _serviceProvider = serviceProvider;
+            _serviceWrapper = serviceWrapper;
+
+            target.OnFlushEvent.Register(() => _crashInfoService = null);
+
+            VTableBuilder builder = AddInterface(IID_ICLRMAService, validate: false);
+            builder.AddMethod(new AssociateClientDelegate(AssociateClient));
+            builder.AddMethod(new GetThreadDelegate(GetThread));
+            builder.AddMethod(new GetExceptionDelegate(GetException));
+            builder.AddMethod(new GetObjectInspectionDelegate(GetObjectInspection));
+            builder.Complete();
+            // Since this wrapper is only created through a ServiceWrapper factory, no AddRef() is needed.
+        }
+
+        protected override void Destroy()
+        {
+            Trace.TraceInformation("ClrmaServiceWrapper.Destroy");
+            _serviceWrapper.RemoveServiceWrapper(ClrmaServiceWrapper.IID_ICLRMAService);
+            _crashInfoService = null;
+        }
+
+        private int AssociateClient(
+            IntPtr self,
+            IntPtr punk)
+        {
+            // If the crash info service doesn't exist, then tell Watson/CLRMA to go on to the next provider
+            return CrashInfoService is null ? HResult.E_NOINTERFACE : HResult.S_OK;
+        }
+
+        private int GetThread(
+            IntPtr self,
+            uint osThreadId,
+            out IntPtr clrmaClrThread)
+        {
+            clrmaClrThread = IntPtr.Zero;
+            if (CrashInfoService is null)
+            {
+                return HResult.E_FAIL;
```

Comment:
The only place that returns E_NOINTERFACE is AssociateClient and it make sense there.

I guess to be more consistent between GetThread and GetException is to return E_INVALIDARG, but it doesn't really matter as long as it fails.  The CrashInfoService is null path in both of them is kind of a internal error since the C++ managed analysis code should never call it with that condition. AssociateClient failing tells the C++ that there is no C# CLRMA support (which is just Native AOT).